### PR TITLE
Fix `DeprecationWarnings` found when running py.test in Python 2.7 with the -3 flag.

### PR DESCRIPTION
**Pull request by rowillia**
I am investigating a migration to Python 3, and to facilitate this we are using the -3 flag as decribed here:
https://docs.python.org/3/howto/pyporting.html#prevent-compatibility-regressions . When using this flag.  

Running through some of my tests with the `-3` flag in python2.7 however I encountered some errors within py.test itself.  This fixes those errors so we can use py.test in order to identify problems with our code as we transition to Python 3.

### CODE REVIEW

**Comment by The-Compiler** [comment]
I'm curious, why are all those `__hash__` methods needed, i.e. what's the corresponding error?

File: _pytest/_code/source.py [hunk 1]
```diff
@@ -52,6 +52,9 @@ def __eq__(self, other):
                 return str(self) == other
             return False
 
+    def __hash__(self):
```
Comment (RonnyPfannschmidt) [inline _pytest/_code/source.py]: this raises an exception for certain

File: _pytest/python.py [hunk 2]
```diff
@@ -1356,6 +1356,9 @@ def __eq__(self, actual):
             return False
         return all(a == x for a, x in zip(actual, self.expected))
 
+    def __hash__(self):
```
Comment (RonnyPfannschmidt) [inline _pytest/python.py]: this breaks in any case a unhashable datastructure is passed as expected

File: _pytest/python.py [hunk 3]
```diff
@@ -1435,6 +1438,9 @@ def __eq__(self, actual):
         # Return true if the two numbers are within the tolerance.
         return abs(self.expected - actual) <= self.tolerance
 
+    def __hash__(self):
```
Comment (RonnyPfannschmidt) [inline _pytest/python.py]: this might work by chance

File: _pytest/python.py (hunk 2 above)
Comment (The-Compiler) [inline _pytest/python.py]: Why would you pass something other than a float?

File: _pytest/_code/source.py (hunk 1 above)
Comment (The-Compiler) [inline _pytest/_code/source.py]: "I never finish anyth..." :wink:

**Comment by rowillia** [comment]
@The-Compiler @RonnyPfannschmidt Python 3 now requires you to implement `__hash__` if you implement `__eq__`.  See https://docs.python.org/3.6/reference/datamodel.html#object.%5F%5Fhash%5F%5F .  For a quick example:

```py
class JustEq(object):
   def __init__(self, x):
     self.x = x
   def __eq__(self, other):
     return self.x == other.x

class Neither(object):
  def __init__(self, x):
    self.x = x

class HashAndEq(object):
   def __init__(self, x):
     self.x = x
   def __eq__(self, other):
     return self.x == other.x
   def __hash__(self):
     return hash(self.x)

{Neither(1), Neither(2)}  # OK
{HashAndEq(1), HashAndEq(2)}  # OK
{JustEq(1), JustEq(2)}  # No Good in Python 3
```

Without these changes, running tests in python2 with the `-3` flag results in error messages like:

```shell
_pytest/_code/code.py:16: DeprecationWarning: Overriding __eq__ blocks inheritance of __hash__ in 3.x
  class Code(object):
```

File: _pytest/_code/source.py (hunk 1 above)
Comment (rowillia) [inline _pytest/_code/source.py]: @RonnyPfannschmidt @The-Compiler What would you suggest is a better hashable value?

**Comment by RonnyPfannschmidt** [comment]
@rowillia the documentation there also suggests that hash should not be implemented for mutable objects
and it suggests that in case of doubt, it should be overridden by assigning or overriding the parent class

so in this case, the only valid option is to either assign `object.__hash__` to those objects or have a immutable equality key (impossible for some of the objects),

also its possible to just assign `__hash__ = None`

however what we absolutely cannot do is just calling hash on mutable objects and hoping for the best

**Comment by rowillia** [comment]
@RonnyPfannschmidt Right, you know the code better than I do as to what should be hashable and what shouldn't.  Replaced `__hash__` implementations with explicitly setting `__hash__` to `None`.

File: _pytest/python.py (hunk 2 above)
Comment (RonnyPfannschmidt) [inline _pytest/python.py]: @The-Compiler see the doc-strings, sequences are supported to make nice expressions with tuples possible

**Comment by nicoddemus** [comment]
Thanks @rowillia!

**Pull request merged**